### PR TITLE
Issue/83 faultinj

### PR DESCRIPTION
**Pull request by vitalych**

### CODE REVIEW

File: src/s2e/Plugins/OSMonitors/Support/ModuleMap.cpp [hunk 1]
```diff
@@ -296,5 +296,53 @@ const ModuleMap::Export *ModuleMap::getExport(S2EExecutionState *state, uint64_t
     return plgState->getExport(address);
 }
 
+void ModuleMap::handleOpcodeInvocation(S2EExecutionState *state, uint64_t guestDataPtr, uint64_t guestDataSize) {
+    S2E_MODULE_MAP_COMMAND command;
+
+    // TODO: factor these checks out from all plugins
```
Comment (adrianherrera) [inline src/s2e/Plugins/OSMonitors/Support/ModuleMap.cpp]: Hahah yeah I agree :)

File: src/s2e/Plugins/OSMonitors/Support/ModuleMap.h [hunk 2]
```diff
@@ -19,10 +20,37 @@
 namespace s2e {
 namespace plugins {
 
+typedef enum S2E_MODULE_MAP_COMMANDS { GET_MODULE_INFO } S2E_MODULE_MAP_COMMANDS;
```
Comment (adrianherrera) [inline src/s2e/Plugins/OSMonitors/Support/ModuleMap.h]: See my comment in https://github.com/S2E/guest-tools/pull/17 asking if we can remove this code duplication by reusing the header file.

**Comment by adrianherrera** [comment]
All looks good! But I'd really like to remove the module map command struct duplication by sharing the header files (like we do with the other command structs).

File: src/s2e/Plugins/OSMonitors/OSMonitor.h [hunk 3]
```diff
@@ -78,6 +78,12 @@ class OSMonitor : public Plugin {
 
     void dumpUserspaceMemory(S2EExecutionState *state, std::ostream &ss);
 
+    // Indicates whether the monitor is loaded and that calling its APIs is allowed.
+    // This is useful for monitors that rely on a guest kernel driver to provide
+    // information to plugins. Plugins should wait that the monitor is initialized
+    // before calling any APIs. This call can be used together with the onMonitorLoad signal.
+    virtual bool initialized() const = 0;
```
Comment (adrianherrera) [inline src/s2e/Plugins/OSMonitors/OSMonitor.h]: Is there any reason why this is a virtual method? It seems that all of the `OSMonitors` (Raw, Linux and Windows) all have a `bool m_initialized` member and all of the `initialized()` methods seem to be the same (i.e. they just return `m_initialized`).

File: src/s2e/Plugins/OSMonitors/Support/ModuleMap.h (hunk 2 above)
Comment (adrianherrera) [reply src/s2e/Plugins/OSMonitors/Support/ModuleMap.h]: Can you at least add a comment here pointing out the duplication with guest-tools and if one is changed that other must be changed as well.

Comment (vitalych) [reply src/s2e/Plugins/OSMonitors/Support/ModuleMap.h]: Yes. I also created an issue to refactor that.
https://github.com/S2E/s2e-env/issues/96

File: src/s2e/Plugins/OSMonitors/OSMonitor.h (hunk 3 above)
Comment (vitalych) [reply src/s2e/Plugins/OSMonitors/OSMonitor.h]: good point

**Pull request merged**